File: app/src/main/java/com/nikola/jakshic/dagger/di/ViewModelModule.java
package com.nikola.jakshic.dagger.di;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import com.nikola.jakshic.dagger.viewModel.RegionViewModel;
import com.nikola.jakshic.dagger.viewModel.PeerViewModel;
import com.nikola.jakshic.dagger.viewModel.BookmarkViewModel;
import com.nikola.jakshic.dagger.viewModel.CompetitiveViewModel;
import com.nikola.jakshic.dagger.viewModel.DetailViewModel;
import com.nikola.jakshic.dagger.viewModel.HeroViewModel;
import com.nikola.jakshic.dagger.viewModel.MatchDetailViewModel;
import com.nikola.jakshic.dagger.viewModel.MatchViewModel;
import com.nikola.jakshic.dagger.viewModel.SearchViewModel;
import com.nikola.jakshic.dagger.viewModel.DaggerViewModelFactory;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
public interface ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(MatchViewModel.class)
    ViewModel bindMatchFragmentViewModel(MatchViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(HeroViewModel.class)
    ViewModel bindHeroFragmentViewModel(HeroViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(SearchViewModel.class)
    ViewModel bindPlayerViewModel(SearchViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(DetailViewModel.class)
    ViewModel bindDetailViewModel(DetailViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(BookmarkViewModel.class)
    ViewModel bindBookmarkViewModel(BookmarkViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(MatchDetailViewModel.class)
    ViewModel bindMatchDetailViewModel(MatchDetailViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(PeerViewModel.class)
    ViewModel bindPeerViewModel(PeerViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(CompetitiveViewModel.class)
    ViewModel bindCompetitiveViewModel(CompetitiveViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(RegionViewModel.class)
    ViewModel bindRegionViewModel(RegionViewModel viewModel);

    @Binds
    ViewModelProvider.Factory bindViewModelFactory(DaggerViewModelFactory factory);
}

File: app/src/main/java/com/nikola/jakshic/dagger/di/ViewModelModule.kt
package com.nikola.jakshic.dagger.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.nikola.jakshic.dagger.bookmark.player.BookmarkViewModel
import com.nikola.jakshic.dagger.common.DaggerViewModelFactory
import com.nikola.jakshic.dagger.competitive.CompetitiveViewModel
import com.nikola.jakshic.dagger.leaderboard.RegionViewModel
import com.nikola.jakshic.dagger.matchstats.MatchStatsViewModel
import com.nikola.jakshic.dagger.profile.ProfileViewModel
import com.nikola.jakshic.dagger.profile.heroes.HeroViewModel
import com.nikola.jakshic.dagger.profile.matches.MatchViewModel
import com.nikola.jakshic.dagger.profile.matches.byhero.MatchesByHeroViewModel
import com.nikola.jakshic.dagger.profile.peers.PeerViewModel
import com.nikola.jakshic.dagger.search.SearchViewModel
import com.nikola.jakshic.dagger.stream.StreamViewModel
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
interface ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(MatchViewModel::class)
    fun bindMatchFragmentViewModel(viewModel: MatchViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(HeroViewModel::class)
    fun bindHeroFragmentViewModel(viewModel: HeroViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(SearchViewModel::class)
    fun bindPlayerViewModel(viewModel: SearchViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(BookmarkViewModel::class)
    fun bindBookmarkViewModel(viewModel: BookmarkViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(MatchStatsViewModel::class)
    fun bindMatchDetailViewModel(viewModel: MatchStatsViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(PeerViewModel::class)
    fun bindPeerViewModel(viewModel: PeerViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(CompetitiveViewModel::class)
    fun bindCompetitiveViewModel(viewModel: CompetitiveViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(RegionViewModel::class)
    fun bindRegionViewModel(viewModel: RegionViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(ProfileViewModel::class)
    fun bindProfileViewModel(viewModel: ProfileViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(MatchesByHeroViewModel::class)
    fun bindMatchesByHeroViewModel(viewModel: MatchesByHeroViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(StreamViewModel::class)
    fun bindStreamViewModel(viewModel: StreamViewModel): ViewModel

    @Binds
    fun bindViewModelFactory(factory: DaggerViewModelFactory): ViewModelProvider.Factory
}